scrapy_autoextract/providers.py

```diff
@@ -45,17 +51,88 @@ def __init__(
         self.stats = stats
         self.settings = settings
 
+    def __before__(self):
+        self.providers.append(self)
+
     async def __call__(self):
         """Make an AutoExtract request and build a Page Input of provided class
         based on API response data.
         """
-        page_type = self.get_page_type()
-        self.stats.inc_value(f"autoextract/{page_type}/total")
+        try:
+            if self.providers[-1] is self:
+                await self._fetch_data()
+        except IndexError:
+            # FIXME: it's not possible to use this alone yet
+            raise Exception("AutoExtractHTMLData should not be used alone")
+
+        data = await self._await_data()
+
+        if "error" in data:
+            self.stats.inc_value("autoextract/error/query")
+            raise QueryError(data["query"], data["error"])
+
+        return self.provided_class(data=data)
+
+    @classmethod
+    def register(cls):
+        """Register this provider for its provided class on scrapy-poet
+        registry. This will make it possible to declare provided class as
+        a callback dependency when writing Scrapy spiders.
+        """
+        register(cls, cls.provided_class)
+
+    @classmethod
+    def get_page_type(cls) -> str:
+        """Page type is defined by the class attribute `item_key` available on
+        `autoextract_poet.page_inputs` classes.
+        """
+        return cls.provided_class.item_key
+
```

Comment:
In the following lines, you'll find some helper properties/setters to make use of Scrapy's request as a centralized storage for multiple providers. We can replace it with a thread-local based storage or perhaps refactor it into something more intelligible. I'm open to suggestions.

Reply:
Why not using `meta`?

Reply:
Yeah, it's possible. Also thinking here that those properties kinda hide what we are doing. We could be using request meta directly.